### PR TITLE
Incorporate Add Bank Account / KYC checks into Send Money + Pay request flow

### PR DESCRIPTION
**Pull request by marcaaron**
cc @nickmurray47 

### Details

Tracking issue and additional context: https://github.com/Expensify/Expensify/issues/185320

This PR adds a way to effectively "block" a specific user action that requires a gold wallet and then to "continue" that action once they successfully make it through these flows. Since the flows all happen on different pages/screens it is tricky to enable the "continue" part of whatever user action they were taking. By using a "pay wall" style component we can save a reference to the component whenever it is mounted and call it's defined "continue" callback.

- [x] Holding on https://github.com/Expensify/App/pull/6992 as the `AdditionalDetailsStep` was recently broken and the test steps here can't be completed until this is fixed.

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/187804
$ https://github.com/Expensify/Expensify/issues/187803

### Tests

1. Start with an account that has a SILVER wallet
2. Attempt to Send Money to someone else
3. Verify you are asked to add a payment method with a popover
4. Select bank account and verify you are brought to the Plaid flow and go through it
5. On success verify, you are next brought to the KYC flow and follow the second set of instructions here which explain how to test the ActivateWallet flow on dev -> https://stackoverflow.com/c/expensify/questions/10607
6. Once you successfully get through this flow the original action you wanted to take (send money) should succeed
7. Repeat with a new account but this time test paying a request
8. Logout after adding a bank account + passing KYC checks then log back in and verify that you do not hit the KYC wall anymore.

### QA Steps
Internal QA

### Tested On

- [x] Web
- [ ] Mobile Web
- [ ] Desktop
- [x] iOS
- [x] Android

### CODE REVIEW

**Comment by marcaaron** [comment]
I'm at an interesting spot with this stuff. A few things are becoming clear...

- These two issues are basically the same with the main difference being that when we push one of the 3 screens (enable payments, add bank, add debit) we'll want it to go on top of the correct stack that we're already on:
    - https://github.com/Expensify/Expensify/issues/187804
    - https://github.com/Expensify/Expensify/issues/187803 
- I think that means we'll need unique routes for all of this stuff and need to pass these in to the `SettlementButton`
- The logic for the `SettlementButton` ideally should work independent of the `SettlementButton` itself so it can also work with the [Transfer Balance button](https://github.com/Expensify/Expensify/issues/187802). @parasharrajat is working on that over in https://github.com/Expensify/App/pull/4177

- There's another tricky part of this to handle which is basically... what happens when we successfully do any of these and need to return to the previous screen and trigger the original event? The possibilities are...
     - add a debit card -> move the user to KYC flow
     - add a bank account -> move the user to KYC flow
     - enable payments -> continue sending money, transferring balance, or paying request
- Now the fun part... each time we do one of those we will dismiss the page by calling `goBack()` or navigating away on success or failure.
- However, the screen underneath will need some logic to _continue_ whatever the user was trying to do before.

I'm unsure the best way to achieve the "continue" logic just yet, but have a few of ideas...

1. Navigate back to the previous page with some params like `{previousRoute: 'bankAccount', success: true}` then wait for these route params in `componentDidUpdate()`, clear them out, and trigger an `onContinue()` callback which will basically perform the original action again and route us to the next place.

2. Use some kind of subscription to trigger an event and subscribe to the event where we place the popover then call the `onContinue()` callback. Instead of "navigating with params" we will `navigate()` or `goBack()` and _also_ publish an event like `EVENT.CONTINUE_PAYMENTS_SETUP`.

3. Similar to `2` we can set a global `onEnablePaymentsContinueCallback()` when our component mounts and clear it out when the component unmounts. Then we can simply call this callback whenever we successfully complete any of the three actions.

4. Onyx based solution. Set some params for the success or failure of any of the actions we tried to take after they are successful or unsuccessful. In the previous screen check to see if these params did not exist before but are now updated in `componentDidUpdate()`. Handle them once then clear them out.

5. Navigate to these routes, but use the screen name and params instead of the `linkTo()` method. This should allow us to [pass a callback as a param](https://stackoverflow.com/questions/44223727/react-navigation-goback-and-update-parent-state) when navigating.

The last option seems to be the easiest so that's what I'll try first.

**Comment by marcaaron** [comment]
Solution `3` ended up being the one that makes the most sense to me.

**Comment by marcaaron** [comment]
Found a better solution than everything else that I thought of which is to create a "pay wall" type component where if the user has not completed checks or added payment method we force them to complete the flow and then "continue" whatever the original action is after. The component uses a ref similar to how the `Growl` component/interface works.

**Comment by marcaaron** [comment]
This is on HOLD still but taking off draft to get some initial thoughts on the overall implementation etc.

**Comment by marcaaron** [comment]
Off hold

**Comment by nickmurray47** [comment]
this looks great thanks for the tag! I need to CP some a PR related to the Bulk Payroll project right but planning on doing a full review on this PR later today.

**Comment by marcaaron** [comment]
Bump

File: src/components/KYCWall.js
Comment (nickmurray47) [inline src/components/KYCWall.js:35]: great choice in class naming

File: src/components/KYCWall.js
Comment (nickmurray47) [inline src/components/KYCWall.js:33]: nab: `originally`

File: src/libs/actions/PaymentMethods.js
Comment (nickmurray47) [inline src/libs/actions/PaymentMethods.js:27]: Which view are we closing here?

File: src/components/SettlementButton.js
Comment (nickmurray47) [inline src/components/SettlementButton.js:127]: side question that is NAB: is it possible to combine these three routes into one? Like we have one component `EnablePayments` that renders a different view depending on which payment method a user is adding? 

Might be an issue of properly separating concerns and not combining flows like we've talked about before.

File: src/components/KYCWall.js
Comment (nickmurray47) [inline src/components/KYCWall.js:71]: why the hardcode? should these be CONSTs?

**Review by nickmurray47 — DISMISSED** [review]
Looks great and tests well, big fan of this solution. All my comments are NABs I think.

**Comment by nickmurray47** [comment]
Before merging @marcaaron, also what is the expectation of how this flow should work when someone already has a `GOLD` wallet? They shouldn't have to add a bank account, right?

Screen Shot 2022-01-06 at 1 46 28 PM

File: src/components/KYCWall.js
Comment (marcaaron) [reply src/components/KYCWall.js:33]: not a NAB typos=bad haha thanks!

File: src/components/KYCWall.js
Comment (marcaaron) [reply src/components/KYCWall.js:71]: Ah hmm, I guess they are hardcoded because I couldn't think of a better way to position this and forgot to go back and improve this so thanks for catching it!

I think maybe the more appropriate thing would be to set the bottom of the popover to some distance away from the top of the button and then set a constant for the overall height of the button. Feels like we did something similar here for the emoji picker.

https://github.com/Expensify/App/blob/021ad5d5b057fcecd98ee70b6bf52938cca5fe32/src/pages/home/report/ReportActionCompose.js#L665-L668

File: src/components/SettlementButton.js
Comment (marcaaron) [reply src/components/SettlementButton.js:127]: Anything is possible I guess. :) What are the advantages of doing things that way?

Comment (marcaaron) [reply src/components/SettlementButton.js:127]: Off the bat, I think we'd need to tell that component which view to render somehow with maybe a route param? 

So the route would end up something like `/enable-payments?subpage=activate-wallet` or something? Feels like that would enable one route to be the "parent" route and we wouldn't need to set up so many routes and less boilerplate. But at the same time we'd have kind of an uglier link to that particular screen.

File: src/libs/actions/PaymentMethods.js
Comment (marcaaron) [reply src/libs/actions/PaymentMethods.js:27]: I can clarify this in the comment. But in this case it will either be the add debit card, add bank account, or enable payments screens.

File: src/components/KYCWall.js
Comment (marcaaron) [reply src/components/KYCWall.js:35]: Thanks!

**Comment by marcaaron** [comment]
>Before merging @marcaaron, also what is the expectation of how this flow should work when someone already has a GOLD wallet? They shouldn't have to add a bank account, right?

No, you **must** add a bank account or debit card in order to do any of these things otherwise you can't pay with the wallet.

File: src/components/KYCWall.js
Comment (marcaaron) [reply src/components/KYCWall.js:71]: Oh hmm it only will give us bottom and right actually:


 
The implementation is pretty complicated for how we calculate these for the emojis eep actually since we

1. Measure the vertical/horizontal placement of the button
2. Re-measure when the window changes
3.  Use that + the size of the picker to display it correctly

Gonna take me a second to create that same behavior here.

Comment (marcaaron) [reply src/components/KYCWall.js:71]: Actually might do it in a follow up. It's gonna be too much to do right now.

Comment (marcaaron) [reply src/components/KYCWall.js]: Fixed these up with CONST for now.

**Comment by nickmurray47** [comment]
cool, just wanted to make sure

File: src/components/KYCWall.js
Comment (marcaaron) [reply src/components/KYCWall.js]: Gonna tackle here https://github.com/Expensify/Expensify/issues/191203

File: src/components/SettlementButton.js
Comment (nickmurray47) [reply src/components/SettlementButton.js:127]: > What are the advantages of doing things that way?

> Feels like that would enable one route to be the "parent" route and we wouldn't need to set up so many routes and less boilerplate. 

Yeah, exactly my thoughts. Trade-off is that the parent component would probably have a lot of different logic to handle the payment methods, which would be a bit messier.

Comment (nickmurray47) [reply src/components/SettlementButton.js:127]: Probably can be left as an improvement.

**Review by nickmurray47 — DISMISSED** [review]
👍 will leave to you @danieldoglas

File: src/components/SettlementButton.js
Comment (marcaaron) [reply src/components/SettlementButton.js:127]: Yea, another way to look at this is that these things don't really have a _ton_ to do with each other and can all happen in isolation. They are perfectly independent modular parts that are only related in that we _sometimes_ need them to run in a specific sequence. 

So, in all a case where you only need one - you end up having to pull in _all_ of them and still have to define which one you need anyway. Honestly, I haven't found a perfect rule for when to keep things portable or smash them together (so I usually lean towards portability when I can). It's a bit of thinking about how something may be used in the future, not obsessing over it too much, but also trying to get it right the first time (because you might not get a chance to refactor later).

**Comment by danieldoglas** [comment]
I'm reviewing and testing this. Not really used to this flow so taking some time.

**Comment by marcaaron** [comment]
Gonna also do some quick tests on the platforms. Just realized this is adding new UI and so should be tested everywhere and have screenshots.

**Comment by marcaaron** [comment]
@danieldoglas please let me know if there are any questions!

File: src/libs/Navigation/AppNavigator/ModalStackNavigators.js
Comment (danieldoglas) [inline src/libs/Navigation/AppNavigator/ModalStackNavigators.js:129]: As discussed with @marcaaron , I think we're adding some complexity in the code duplicating the `AddDebitCardPage` and `AddPersonalBankAccountPage` in two stacks.

If we duplicate them here, we need to pass them as parameters in `IOUConfirmationList`, `SettlementButtton` and `IOUDetailsModal` instead of just using them in `KYCWall`. 

If we can simplify it using only two routes, that would help us in maintenance in the future.

What do you think @nickmurray47 ?

Also, this is NAB.

File: src/CONST.js
Comment (danieldoglas) [inline src/CONST.js:348]: Do we want to work with fixed values here independent of the platform and screen size? Just to be sure this will be how we're imagining in every device.

**Review by danieldoglas — COMMENTED** [review]
Leaving this as a comment so we can discuss my two comments before approving, but overall LGTM

File: src/CONST.js
Comment (marcaaron) [reply src/CONST.js:348]: This is a bit of temp code and a follow up created to address this [here](https://github.com/Expensify/Expensify/issues/191203)

**Comment by marcaaron** [comment]
When testing on iOS seems like the an error appears...



Kind of confused because I have not added and `ScrollView` in this PR so going to check against `main` to see if the issue exists there.

**Comment by marcaaron** [comment]
Hmm yes the VirtualizedList warning exists on `main` as well. We should maybe create an issue to look into this - but can ignore it as we are still under beta and it wasn't introduced in this PR.

File: src/libs/Navigation/AppNavigator/ModalStackNavigators.js
Comment (marcaaron) [reply src/libs/Navigation/AppNavigator/ModalStackNavigators.js:129]: I like the idea of using generic routes for these then they will exist outside of the underlying stack. 

But I think in that case we should not have both `<` and `x` in their headers because they would both do the same thing.

This wasn't really ironed out in the doc so I followed the existing convention (we already had `IOU_Enable_Payments` set up with a unique route).

**Comment by marcaaron** [comment]
Ah I found a bug with some of the logic while testing and updated it. If you add a bank account and then log out and then try to send an IOU again after logging in we will not check the most recent list of payment methods and end up asking the user to enter one again.

@nickmurray47 mind re-reviewing?

**Comment by nickmurray47** [comment]
Yeah, can take another look! catching up on context right now

File: src/components/SettlementButton.js
Comment (nickmurray47) [reply src/components/SettlementButton.js:127]: Yes, I also a much bigger fan of keeping things separate and portable to begin with because I think it's easier to combine things later in a refactor than pull disparate components out of a single one.

**Comment by marcaaron** [comment]
The ScrollView thing is being handled here https://github.com/Expensify/App/pull/7088

**Review by nickmurray47 — APPROVED** [review]
Nice, LGTM! Thanks for adding the screenshots in for the other platforms, I should've asked for those in the beginning. 

Do you also mind adding just one last step to QA:
`8. Logout after adding a bank account, log back in and verify that you do not hit the pay wall`

**Review by danieldoglas — APPROVED** [review]
LGTM

**Comment by mvtglobally** [comment]
@marcaaron @danieldoglas  @nickmurray47  PR is ready to be QA-ed internally

**Comment by marcaaron** [comment]
@mvtglobally This can be checked off. It's behind a beta and has been tested on staging.

**Pull request merged**
